tardis/visualization/tools/convergence_plot.py

```diff
@@ -0,0 +1,439 @@
+
+            # the next three for emitted, absorbed and requested luminosities
+            for index, luminosity in zip(range(1, 4), self.luminosities):
+                self.t_inner_luminosities_plot.data[index].x = x
+                self.t_inner_luminosities_plot.data[index].y = self.value_data[
+                    luminosity
+                ]
+                self.t_inner_luminosities_plot.data[index].hovertemplate = (
+                    "<b>%{y:.4g}</b>" + "<br>at X = %{x}<br>"
+                )
+
+            # last is for the residual luminosity
+            y = [
+                ((emitted - requested) * 100) / requested
+                for emitted, requested in zip(
+                    self.value_data["Emitted"], self.value_data["Requested"]
+                )
+            ]
+
+            self.t_inner_luminosities_plot.data[4].x = x
+            self.t_inner_luminosities_plot.data[4].y = y
+            self.t_inner_luminosities_plot.data[
+                4
+            ].hovertemplate = "<b>%{y:.2f}%</b> at X = %{x:,.0f}"
+
+    def update(self, export_cplots=False, last=False):
+        """
+        Update the convergence plots every iteration.
+
+        Parameters
+        ----------
+        export_cplots : bool
+            Displays the convergence plots again using plotly's ``notebook_connected`` renderer.
+            This helps display the plots in notebooks when shared on platforms like nbviewer.
+            Please see https://plotly.com/python/renderers/ for more information.
+            Defaults to False.
+        last : bool
+            True if it's last iteration.
+        """
+        if self.iterable_data != {}:
+            # build only at first iteration
+            if self.current_iteration == 1:
+                self.build()
+
+            self.update_plasma_plots()
+            self.update_t_inner_luminosities_plot()
+
+            # data property for plasma plots needs to be
+            # updated after the last iteration because new traces have been added
+            if hasattr(self, "plasma_plot_config") and last:
+                if "data" in self.plasma_plot_config:
+                    self.override_plot_parameters(
+                        self.plasma_plot, self.plasma_plot_config
+                    )
+
+        self.current_iteration += 1
+
+        # the display function expects a Widget, while
+        # fig.show() returns None, which causes the TraitError.
+        if export_cplots:
+            with suppress(TraitError):
+                display(
+                    widgets.VBox(
+                        [
+                            self.plasma_plot.show(
+                                renderer="notebook_connected"
+                            ),
+                            self.t_inner_luminosities_plot.show(
+                                renderer="notebook_connected"
+                            ),
```

Comment:
Is it possible to not call show instead just return FigureWidget - I think then display will just show it and you won't need to supress trait error?

Reply:
Actually, this exact thing happens in the `build` function. The plots show up locally without having to suppress any errors, but in order to display them on platforms like nbviewer, we need to have the [`notebook_connected`](https://plotly.com/python/renderers/) renderer. Unfortunately `fig.show()` returns None which causes the error we need to suppress. 

If it would have worked, we would have not needed the `export_cplots` option but would have seen the plots two times online as we see now locally, if we had run the simulation with `export_cplots=True`. One due to the build function and the other due to the `notebook_connected` renderer. The build function fails to display the plots online actually.

Reply:
OK - thanks for explanation!